Proposed Restructuring of Orleans.Template NuGet Packages

jthelin opened the pull request:
This is a proposal to restructure the Orleans NuGet packages ```Microsoft.Orleans.Templates.Microsoft.Interfaces``` and ```Microsoft.Orleans.Templates.Microsoft.Grains``` to address several outstanding issues / bugs with the v1.0.3 NuGet packages currently.

Issues to be addressed:

1. PR #229 demonstrates that 1.0.3 Orleans.Templates packages **do not work at all** with apps using the new recommended way to do "Automatic Package Restore" with NuGet 2.8 [as used by current VS 2013 Update 4]
http://docs.nuget.org/consume/package-restore
http://docs.nuget.org/Consume/Package-Restore/Migrating-to-Automatic-Package-Restore
2. PR #226 is a change which attempts to apply a kludgy workaround to get Orleans CodeGen to work in some repo scenarios by introducing additional Referenced copies of Orleans.dll. However, that creates a non-deterministic version race condition regarding which copy of Orleans.dll will actually get used, as i explain in the comments on PR #226. This change addresses the underlying problem of getting codegen.exe to work correctly in the build-time environment through using the package /tools/ directory for what it was intended for - executables to run at build time.
3, The Orleans codegen tool is only required at build time, so adding extra References to that .exe into every application project creates extra bloat in the app binaries that people will deploy.
4. There are **13** different Orlerans NuGet packages currently, and this change allows us to get rid of 2 of those without any loss of functionality or packaging fidelity. :)

Extra notes:

- The main 'meat' of this change is in commit cf636e6
- This change includes the fix from PR #228 because it is required here too, but once that PR is merged into master I can rebase to master and remove 272e00c
- For illustrative purposes, I also included commit e98fbe9 here which shows the minimal viable delta [Core + Templates pkgs => 1.04, other pkgs remain at 1.0.3] to get the HelloWorldNuget sample app to work again. In practice, i would recommend we always update sample apps consistently to the same version 1.0.4, so consider that part of this change as informative only and the "real" fix would be atomically applied to PR #229 after we merge this PR to master [ minus e98fbe9]
- I also include 0119019 here so that packages are always built against the same versions as themselves. Requiring manual changes to 13 [or 11] different .nuspec files is error prone way to handle the default case. Any special  hard dependency requirements in a hitfix situation can be handled explicitly on a "hitfix" branch when necessary, but automating the normal default case seems highly desirable.
- This change is **only** targeting to address the problems with the Templates.* packages at this time. There are many other things that could be done to consolidate and combine the other packages into a more manageable number, but i do not want to conflate those discussions / decisions with this fix. The other packages.

It you get to this point and feel like you want to ask me to squash this change into a single commit, then go grab a second coffee and re-read the above! ;)

sergeybykov commented on src/Build/Version.txt:
```diff
@@ -1 +1 @@
-1.0.3
```
Need to bump it to 1.0.5. We already used a private 1.0.4 internally for a major code migration.

sergeybykov commented on src/NuGet/Microsoft.Orleans.Client.nuspec:
```diff
@@ -18,9 +18,9 @@
-      <dependency id="Microsoft.Orleans.Core" version="1.0.3" />
-      <dependency id="Microsoft.Orleans.OrleansAzureUtils" version="1.0.3" />
-      <dependency id="Microsoft.Orleans.OrleansProviders" version="1.0.3" />
+      <dependency id="Microsoft.Orleans.Core" version="$version$" />
```
Does this mean we won't be able to update individual NuGets anymore, as they will require all of them updated at once? That would defeat one of the purposes of the fine-grain NuGets.

veikkoeeva replied:
I'll write this a bit more open, if someone else is also wondering...

It looks like this means the client is required always to have the same version of 
* Microsoft.Orleans.Core
* Microsoft.Orleans.OrleansAzureUtils
* Microsoft.Orleans.OrleansProviders

I.e. it's dependencies. This would mean, I gather, those libraries need to updated in sync too. Would this always be the case? Taken from the library names alone, it looks like they could be released individually with differing version numbers (esp. with relation to sematic versioning). Though the reality may be currently they are always released in sync and the having the same version number looks more clear for an outsider.

How to maintain separate packages?

sergeybykov replied:
I think we should keep specific minimum versions as before, i.e. 1.0.5 in this case, so that we can publish non-breaking changes to a package as, say, 1.0.6 and not force everyone to update all other packages.

jthelin replied:
Well, i think we can *guess* about backward compatibility, but can't be 100% sure unless it is explicitly tested. [Not something anyone is planning to do any time soon, AFAIK]

So this would mean >= 1.0.5, which would be the new versions.
``` xml
<dependency id="Microsoft.Orleans.Core" version="1.05" />
```

Are you actually saying that we should leave this at ```1.0.3``` instead?

sergeybykov replied:
I'm suggesting to replace 1.0.3 with 1.0.5 in the dependency nodes because this one is a breaking change, but not put $version$ there, so that we don't automatically bump the minimum required version of dependencies when we update one or more of the packages. For example, we can do several iterations of the Templates.* packages without updating any of the packages with binaries, especially if we keep the Development package and do not directly put binaries into Templates.*.

jthelin replied:
OK, i got the first part.

I don't think keeping the Development package is viable though, for reasons i just posted below.

sergeybykov commented on src/NuGet/Microsoft.Orleans.Core.nuspec:
```diff
@@ -24,6 +24,5 @@
   </metadata>
   <files>
     <file src="Orleans.dll" target="lib\net45" />
-    <file src="Orleans.xml" target="lib\net45" />
```
Why remove XML docs?

jthelin replied:
I never quite know how useful the xml codedocs are without the .pdb files too, and/or whether the xml code-docs are supposed to be in the "symbol package" or here?

http://docs.nuget.org/Create/Creating-and-Publishing-a-Symbol-Package

jthelin replied:
I'll add back the .xml files until we can clarify requirements for this stuff.

sergeybykov commented on src/NuGet/Microsoft.Orleans.Development.nuspec:
```diff
@@ -1,34 +0,0 @@
-﻿<?xml version="1.0"?>
```
Do you think it's better to duplicate the content of this package in two, and potentially more in the future, other packages?

jthelin replied:
Yes I do. 

There are **only** two packages that need those files for each of the two "developer artifacts types" - one for grains and one for grain-interfaces. 

As you discovered yourself, it is easy to locate the tools .exe files from build .props . targets file IFF they are in the same NuGet package, but extremely hard to do that if they are spread across two NuGet packages. 

This latter point out-ways almost everything else in my mind, especially if you want to retain the flexibility of combining single packages with different versions. We already have a very strong tie between codegen.exe and the core code-gen interfaces in Orleans.dll, so i don't think we can realistically every undo that any time soon. 
For example, PR #194 created a breaking-change in the binary API between GrainReference abstract class and the type-specific generated grain-reference classes, which required matching changes on both Orleans.dll and codegen.exe.

I don't think we are planing to introduce any additional "developer artifact" types other than grin/grain-interfaces any time soon, and even if we do i suspect there will only even be a very small number of additions to deal with so i don't think the extra "de-normalization" really buys us anything here significant here.

jthelin replied:
Looked at in hindsight, we actually should have bumped the version number on .Core and .CodeGen packages [only] with PR #194 because of the breaking change to binary API, although we currently don't have any way to do that in build for individual packages -- only all packages.

jthelin commented:
Rebased against latest master branch.
Addressed all the review feedback comments [i think - please check i didn't miss anything ;) ]
Updated package versions to 1.0.5.

jthelin commented:
I also had to change the pkg reference Unofficial.Microsoft.WindowsAzure.ServiceRuntime to 2.4 rather than 2.5 in HelloWorldNuget to avoid "version mismatch" warnings from build.

I believe this is the same issue related to #186 which @gabikliot is currently looking into.

sergeybykov commented:
I was working on your PR in parallel. :-) Take a look at https://github.com/sergeybykov/orleans/commit/121d4d6638b9bd7d654944e993c5039a7e4c5aa2. The key thing I did there is I removed all the dependencies from Core and moved them to Client and Server. This way I was able to eliminate SIX storage-related nugets from both interfaces and grains projects.

I think that's what we want - only have Orleans.dll referenced in interfaces/grains projects, and only bring the myriad of dependencies to silo or client projects. Looks much slimmer to me now.

jthelin commented:
Ok cool. I'll merge your changes on top of mine  @sergeybykov and push the combined to this PR

jthelin commented:
Merged additions from @sergeybykov 

https://github.com/sergeybykov/orleans/commit/121d4d6638b9bd7d654944e993c5039a7e4c5aa2

jthelin commented:
This was definitely Git Merge 201 class for me! :)

sergeybykov commented:
Would it be better to squash some/all of the commits now, as they change same files more than once?

Do we need to update the VSIX?

jthelin commented:
You are obsessed with squashing commits! ;)
Git is more than able to works everything out by sequentially applying the stream of edits to get us to the end of the branch, even if multiple commits change same file.

Regarding the corresponding VSIX updates, i suggest we get this change branch merged ASAP rather than delaying if there is any comments on another PR for that. 

I already have branch prepared for VSIX changes, but would like to sanity check a little more against this change set before declaring ready.
https://github.com/jthelin/orleans/tree/Vsix-Templates-NuGet

sergeybykov commented:
No, I'm not suggesting to hold this one for the VSIX update. Just trying to get to the next topic. :-) I agree that we need to merge this one ASAP. Just wanted to squash these 1.0.3->1.0.4->1.0.5 changes, so that they are just 1.0.3->1.0.5. :-)

jthelin commented:
Unless i missed making a change to 1.0.5 somewhere?
